Implement Technical Staff Login via External Login Provider

HlibBondarev opened the pull request:
1) Created Moderator entity.
2) Created TechAdmin entity.
3) Configured entity relationships.
4) Updated OutOfSchoolDbContext: added DbSet properties for Moderator and TechAdmin. 
5) Create repositories for new entities (Moderator, TechAdmin) using the generic approach. 
6) Modify existing authentication flow:
- updated SignInUserAsync() method of ExternalAuthController();
- added BuildAdminModeratorClaims() method of  ExternalAuthController. 
7) UI Changes:
- Create new Technical Staff Login Page;
- Style modifications - create new CSS for technical staff components; — Added js-code to handle button clicks.
8) Expanded test coverage to verify the authentication flow for users with technical staff roles. 
9) Removed redundant UserName property from ShortUserDto.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added support for new user roles: Moderator and TechAdmin, including dedicated login options and role validation during authentication.
  - Introduced new database entities and management for Moderator and TechAdmin roles.

- **User Interface**
  - Updated login page to allow switching between provider/employee and technical staff (moderator/techadmin) role selection.
  - Added a "Switch user type" toggle and improved role selection UI.

- **Localization**
  - Enhanced and reorganized English and Ukrainian resource files to include messages for new roles and improve structure.

- **Bug Fixes**
  - Improved error handling with role-specific messages during authentication.

- **Documentation**
  - Added and updated XML documentation for improved code clarity.

- **Tests**
  - Expanded test coverage for new roles and authentication scenarios.

- **Chores**
  - Updated CSS styles for new UI elements and modernized resource file schemas.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HlibBondarev committed:
1) Created Moderator entity.
2) Created TechAdmin entity.
3) Configured entity relationships.
4) Updated OutOfSchoolDbContext: added DbSet properties for Moderator and TechAdmin.
5) Create repositories for new entities (Moderator, TechAdmin) using the generic approach.
6) Modify existing authentication flow:
- updated SignInUserAsync() method of ExternalAuthController();
- added BuildAdminModeratorClaims() method of  ExternalAuthController.
7) UI Changes:
- Create new Technical Staff Login Page;
- Style modifications - create new CSS for technical staff components;
— Added js-code to handle button clicks.
8) Expanded test coverage to verify the authentication flow for users with technical staff roles.
9) Removed redundant UserName property from ShortUserDto.

HlibBondarev commented:
@coderabbitai review

HlibBondarev committed:
Filled in properties IsDeleted and Individual for test Moderator and TechAdmin in the ExternalAuthControllerTests class.

HlibBondarev committed:
Fixed the issue mentioned CodeRabbit - optimized external authentication scheme retrieval.

HlibBondarev commented on OutOfSchool/OutOfSchool.AuthCommon/Views/Auth/Login.cshtml:
```diff
@@ -82,6 +86,70 @@
+        <section class="icon_wrapper" id="for_technical_staff" style="display:none">
+            @{
+                var loginAdminProviders = (await SignInManager.GetExternalAuthenticationSchemesAsync()).ToList();
+                if (loginAdminProviders.Count == 0)
```
Ok, fixed

HlibBondarev committed:
1) Fixed tests in the ExternalAuthControllerTests class.
2) Renamed BuildAdminModeratorClaims() method to BuildTechnicalStaffClaims() in the ExternalAuthController.

HlibBondarev committed:
Fixed the issue mentioned CodeRabbit for js-code in login.js:
- Reduce code duplication in role selection handlers;
- Improve the user type toggle implementation.

HlibBondarev committed:
Added an abstract class - Technical Staff. Inherited Moderator and TechAdmin entities from this new class.

HlibBondarev committed:
Fixed the issues mentioned SonarCloud for login.cshtml:
- Added an 'aria-label' attribute to forms.
- Removed attribute 'role' from forms.
- Wrapped input element with label.

coderabbitai[bot] committed:
📝 Add docstrings to `HlibBondarev/Implement_Technical_Staff_Login` (#1836)

Docstrings generation was requested by @HlibBondarev.

* https://github.com/ita-social-projects/OoS-Backend/pull/1835#issuecomment-2885005069

The following files were modified:

* `OutOfSchool/OutOfSchool.AuthCommon/Controllers/ExternalAuthController.cs`
* `OutOfSchool/OutOfSchool.BusinessLogic/Models/ShortUserDto.cs`
* `OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/ModeratorConfiguration.cs`
* `OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/TechAdminConfiguration.cs`
* `OutOfSchool/OutOfSchool.DataAccess/OutOfSchoolDbContext.cs`

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Co-authored-by: Hlib Bondarev <Hlib.Bondarev@gmail.com>

HlibBondarev committed:
Removed a blank line in the SignInUserAsync() method of the ExternalAuthController class.

DmyMi committed:
updated UI

DmyMi commented on OutOfSchool/OutOfSchool.AuthCommon/Controllers/ExternalAuthController.cs:
```diff
@@ -31,7 +32,12 @@ public class ExternalAuthController : Controller
     private readonly OutOfSchoolDbContext dbContext;
     private readonly IAikomProviderService aikomProviderService;
     private readonly OpenIddictClientService openIddictClientService;
+    private readonly IEntityRepositorySoftDeleted<Guid, Moderator> moderatorRepository;
```
we have a sprcial `ISensitiveEntityRepositorySoftDeleted` for these types of Guid keyed entities

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.AuthCommon/Controllers/ExternalAuthController.cs:
```diff
@@ -144,19 +161,27 @@ private async Task<IActionResult> SignInUserAsync(
-                return await GetErrorMessageResult(result, localizer["IndividualOrProviderNotFound"]);
+                var errorMessage = selectedRole switch
+                {
+                    _ when Role.Provider.ToString().Equals(selectedRole, StringComparison.OrdinalIgnoreCase) => "IndividualOrProviderNotFound",
```
i think the modern approach is to use `nameof(Role.Provider).Equals(...)` 😅

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.AuthCommon/Controllers/ExternalAuthController.cs:
```diff
@@ -192,9 +216,31 @@ private async Task<IActionResult> SignInUserAsync(
             }
+            else if (Role.TechAdmin.ToString().Equals(selectedRole, StringComparison.OrdinalIgnoreCase))
+            {
+                if ((await techAdminRepository.GetById(individual.Id)) is null)
```
`dbContext.SomeEntity.Any(b => b.Id == someId)` is transformed into an `EXISTS` query to database. This does not return any data, so more optimal in checking if the record exists. Technically, our get by id can look into change tracker and return the entity if it is stored there instead of going to DB, but as logins happen once per day - i'm sure the entity will not be there and EF will query the database.

So check if we have an `Exists` method in our repositories

HlibBondarev replied:
Ok, 
"Exists" is absent, I used "Any" instead of it

DmyMi commented on OutOfSchool/OutOfSchool.AuthCommon/Controllers/ExternalAuthController.cs:
```diff
@@ -144,19 +161,27 @@ private async Task<IActionResult> SignInUserAsync(
-            List<Claim> claims = [];
-            
-            if (Role.Provider.ToString().Equals(selectedRole, StringComparison.OrdinalIgnoreCase) || 
+            List<Claim> claims;
```
Leave the `= []` initialiser. It helps with NPE if for what ever reason we change some logic that uses this claims array before `Build...Claims` methods are called. Empty list memory footprint is ~32 bytes which is negligible.

HlibBondarev replied:
Ok

DmyMi commented on OutOfSchool/OutOfSchool.AuthCommon/Controllers/ExternalAuthController.cs:
```diff
@@ -404,22 +450,46 @@ private List<Claim> BuildProviderClaims(
+    /// <param name="userInfo">External user information retrieved from the authentication provider.</param>
+    /// <param name="result">The authentication result containing selected role information.</param>
+    /// <returns>A list of claims representing the user's identity and role.</returns>
+    private static List<Claim> BuildTechnicalStaffClaims(
```
This comment is not directly about this method, but this is only the first part of authentication. Later these claims are used to generate new Principal for JWT Token. Please check the usage of `OutOfSchool.AuthorizationServer/Services/ProfileService.cs` and its method `GetAdditionalClaimsForRoleAsync`. Double check that technical staff claims are correctly set to the token (as i don't remember if i made that method only for providers 😅 )

HlibBondarev replied:
It's all right)

HlibBondarev committed:
Fixed the issues mentioned by @Dmytro Minochkin with ExternalAuthController.

DmyMi committed:
Updated role selection logic in the login.js file to unify handling of roles under a single input.

HlibBondarev committed:
1) Fixed login.js - removed the redundant arg in the function setupRoleSelection(), used the const instead of it.
2) Removed an empty line from the Moderator class.
3) Added migrarion - 20250523092435_AddModeratorAndTechAdmin.
